devlib/instrument/baylibre_acme.py

```diff
@@ -0,0 +1,557 @@
+                       'as detected probes ({}).')
+                raise ValueError(msg.format(len(probe_names),
+                                            len(self.iio_context.devices)))
+
+        probes = [IIOINA226Instrument(d) for d in self.iio_context.devices]
+
+        self.probes = (dict(zip(probe_names, probes)) if probe_names
+                       else {p.iio_device.id : p for p in probes})
+        self.active_probes = set()
+
+        for probe in self.probes:
+            for measure in ['voltage', 'power', 'current']:
+                self.add_channel(site=probe, measure=measure)
+        self.add_channel('timestamp', 'time_us')
+
+        self.data = pd.DataFrame()
+
+    def check_version(self):
+        msg = ('The IIO drivers running on {} ({}) are out-of-date; '
+               'devlib requires {} or later.')
+
+        if iio.version[:2] < self.MINIMAL_HOST_IIO_DRIVERS_VERSION:
+            ver_str = '.'.join(map(str, iio.version[:2]))
+            min_str = '.'.join(map(str, self.MINIMAL_HOST_IIO_DRIVERS_VERSION))
+            raise HostError(msg.format('this host', ver_str, min_str))
+
+        if self.version[:2] < self.MINIMAL_ACME_IIO_DRIVERS_VERSION:
+            ver_str = '.'.join(map(str, self.version[:2]))
+            min_str = '.'.join(map(str, self.MINIMAL_ACME_IIO_DRIVERS_VERSION))
+            raise TargetError(msg.format('the BBB', ver_str, min_str))
+
+    # properties
+
+    def probes_unique_property(self, property_name):
+        probes = self.active_probes or self.probes
+        try:
+            # This will fail if there is not exactly one single value:
+            (value,) = {getattr(self.probes[p], property_name) for p in probes}
+        except ValueError:
+            msg = 'Probes have different values for {}.'
+            raise ValueError(msg.format(property_name) if probes else 'No probe')
+        return value
+
+    @property
+    def version(self):
+        return self.iio_context.version
+
+    @property
+    def OVERSAMPLING_RATIOS_AVAILABLE(self):
+        return self.probes_unique_property('OVERSAMPLING_RATIOS_AVAILABLE')
+
+    @property
+    def INTEGRATION_TIMES_AVAILABLE(self):
+        return self.probes_unique_property('INTEGRATION_TIMES_AVAILABLE')
+
+    @property
+    def sample_rate_hz(self):
+        return self.probes_unique_property('sample_rate_hz')
+
+    @sample_rate_hz.setter
+    # This setter is required for compliance with the inherited methods
+    def sample_rate_hz(self, value):
+        if value is not None:
+            raise AttributeError("can't set attribute")
+
+    # initialization and teardown
+
+    def setup(self, shunt_resistor,
+              integration_time_bus,
+              integration_time_shunt,
+              oversampling_ratio,
+              buffer_samples_count=None,
+              buffer_is_circular=False,
+              absolute_timestamps=False,
+              high_resolution=True):
```

Comment:
I maybe misunderstanding the intent here, but it seems this code should be part of `__init__` (or maybe `reset`?) instead. `setup` is not guaranteed to be called on every use of the instrument -- it's intended for "one-time setup" actions, such as installing executable on the target, and so typically would not be modifying the state of `self`, unless overriding something that would otherwise be set earlier on the assumption that such one-time setup has already been performed (e.g. setting path to the installed executable inside `__init__`).

Reply:
I must admit I struggled to partition the different steps into `__init__`, `setup` and `reset`.

IIO uses a buffer to transfer data between the ACME-BBB and the host. Although required for a transfer to take place, this buffer is created and deleted independently of the IIO context and the establishment of a connection. The properties of the transfer itself are dictated by the properties of the buffer.

Here is my reasoning:
- A constructor (`__init__`) sets an instance to a valid initial state and is obviously not intended to be called more than once on an instance. Therefore, it makes sense to have `__init__` creating the IIO context (and the underlying connection to the BBB) as it will only happen once. On the other hand, because a succession of buffers with varying parameters can be used with the same IIO context, I didn't think creating the buffer in `__init__` made sense. Indeed, in IIO, there is no need for taking the IIO context down and up again just to change the buffer parameters (which we would effectively have to do if the code was in `__init__`);
- [`reset`](https://devlib.readthedocs.io/en/latest/instrumentation.html#Instrument.reset) has a well defined API preventing me to add the extra parameters required by the creation of an IIO buffer;
- The [documentation for `Instrument.setup`](https://devlib.readthedocs.io/en/latest/instrumentation.html#Instrument.setup) states that
    > Parameters this method takes are **particular to subclasses (see documentation for specific instruments below)**. **What actions are performed by this method are also instrument-specific**. Usually these will be things like installing executables, **starting services**, deploying assets, etc. Typically, this method needs to be invoked at most once per reboot of the target (**unless `teardown()` has been called**), **but see documentation for the instrument you’re interested in**.

    which lead me to encapsulate the IIO buffer creation in `setup` (actually, in the current implementation only the IIO buffer setup happens in `setup`, the creation itself is done "inside" `start` so that the buffer only exists when necessary) in, what I thought to be, a way complying with the standard `Instrument`.

The way the flow works is:

1. an IIO context has to be created and its IIO channels are discovered - `__init__`
1. an IIO buffer with a given set of characteristics has to be created - `setup`
1. the IIO channels have to be enabled - `reset` [IIO is now ready to capture]
1. the capture of the selected channels takes place in the given context through the given buffer - `start`-`stop`
1. the IIO buffer is `free`d - `teardown` (actually `stop`)
1. IIO channels are potentially disabled - `teardown`
1. loop to 2. if desired;
1. the IIO context is deleted

And I tried to wrap this IIO flow in the methods `Instrument` provides which, I believed, seemed to work nicely together. Hence the little "calling graph" I joined to the documentation.

Reply:
> I must admit I struggled to partition the different steps into `__init__`, `setup` and `reset`.

The distinction is between when/how often things need to happen:

Some things are fairly persistent - need to only be done once for a particular device (e.g. installing an executable) or a single boot of the device (e.g. starting a daemon/service,  setting up a persistent communications channel, etc). This is what `setup` is intended for. In particular, it may not be called every time an `Instrument` is instantiated. This so that you don't have to keep doing these things, e.g. every time you want to re-run a notebook.

It is expected that a single interactive session (a WA run, notebook/script execution, etc) would have an instance of an `Instrument` persist throughout the session (rather than needed to re-instantiate it each time measurements need to be collected). So anything that needs to be done to configure the `Instrument` for interaction with a particular target should be done inside '__init__'. 

Finally, `reset` is used to prepare the `Instrument` for collecting a particular set(s) of  measurements. This typically just involves specifying which channels (for instruments that support multiple channels) should be collected.

To summarise:

- `setup` is called once per target setup/configuration, and may not be called every time an `Instrument` is created.
- `__init__` is called once per "interactive session" with the target.
- `reset` is called prior to collecting a particular set of measurements.

`setup` was intended to avoid repeating unnecessary steps in situations where target state can be assumed to persist across code invocations. Current applications though, do not make such assumptions, so right now,`setup` will get invoked once per instance of an `Instrument`. I think, for now, current implementation might be OK.